src/charm.py

```diff
@@ -1,103 +1,283 @@
-            container = self.unit.get_container("httpbin")
-            # Verify that we can connect to the Pebble API in the workload container
-            if container.can_connect():
-                # Push an updated layer with the new config
-                container.add_layer("httpbin", self._pebble_layer, combine=True)
-                container.replan()
-
-                logger.debug("Log level for gunicorn changed to '%s'", log_level)
-                self.unit.status = ops.ActiveStatus()
-            else:
-                # We were unable to connect to the Pebble API, so we defer this event
-                event.defer()
-                self.unit.status = ops.WaitingStatus("waiting for Pebble API")
-        else:
-            # In this case, the config option is bad, so block the charm and notify the operator.
-            self.unit.status = ops.BlockedStatus("invalid log level: '{log_level}'")
+        self.framework.observe(self.database.on.database_created, self._on_database_created)
+        self.framework.observe(self.database.on.endpoints_changed, self._on_database_changed)
+
+        self.framework.observe(self.ingress.on.ready, self._on_ingress_ready)
+        self.framework.observe(self.ingress.on.revoked, self._on_ingress_revoked)
+
+        self.framework.observe(
+            self.loki_consumer.on.promtail_digest_error,
+            self._promtail_error,
+        )
 
     @property
-    def _pebble_layer(self):
-        """Return a dictionary representing a Pebble layer."""
-        return {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
+    def _metrics_external_url(self) -> str:
+        return normalise_url(self.ingress.url) if self.ingress.is_ready() else ""
+
+    @property
+    def _pebble_layer(self) -> Layer:
+        pebble_layer: LayerDict = {
+            "summary": "GLAuth Application layer",
+            "description": "pebble layer for GLAuth service",
             "services": {
-                "httpbin": {
+                self._service_name: {
                     "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
-                    "startup": "enabled",
-                    "environment": {
-                        "GUNICORN_CMD_ARGS": f"--log-level {self.model.config['log-level']}"
-                    },
+                    "summary": "GLAuth Operator layer",
+                    "startup": "disabled",
+                    "command": '/bin/sh -c "{} 2>&1 | tee {}"'.format(
+                        self._glauth_service_command,
+                        str(self._log_path),
+                    ),
                 }
             },
         }
+        return Layer(pebble_layer)
+
+    @property
+    def _ldap_port(self) -> str:
+        return "3893"
+
+    @property
+    def _http_port(self) -> str:
+        return "5555"
```

Comment:
Let's store the ports in global variables as we do for other charms.

Reply:
ok